### DISCUSSION TITLE
Out argument on Custom Activity

### DISCUSSION THREAD

**Discussion by Aoshua** [elsa-workflows/elsa-core#893] [Q&A] [answered]
Hello, 

I am very excited to incorporate Elsa into my team's software!

I am trying to convert some existing Activities that I have for WF to Elsa, and I have some questions about in-arguments and out-arguments. My activities often take arguments (in-arguments) and return objects and primitives to be used as arguments in the next activity (out-arguments). If I'm understanding [this tutorial](https://elsa-workflows.github.io/elsa-core/docs/extensibility-custom-activities), I should be able to achieve in-arguments with properties on my activity, but how can I achieve out-arguments? It seems that an `OutcomeResult` can only be a string with statuses such as `Done`, `True` and `False`?

**Comment by tomy2105** [comment]
Actually, what you are looking for is `OutputResult`. And it can be object of any kind.
Your custom activity would returns both `OutcomeResult` and `OutputResult` usually combined, something like `return Outcome("AnOutput ", someObject)`.

**Reply by Aoshua** [reply]
Thanks for your response!

I am not seeing anything in the namespaces called `OutputResult` though I do see a `OutcomeResult` and an `Output`. Here are a few things I've tried:

```
var myObject = new object();
return Outcome("Done", myObject);
return Outcome("Done", new { Param1 = 1 });
return Outcome("Done", new OutcomeResult(new [] { "" }));
// These all tell me that the Outcome() method only takes one argument.
```

```
return OutcomeResult("someKey", someValue);
// This tells me that OutcomeResult is non-invokable & cannot be treated as a method
```

```
Output.SetVariable("someKey", myObject);
// This at least compiles... Could this be it? (about to go test)

**Reply by Aoshua** [reply]
I've noticed that my `ActivityPropertyAttribute` is also missing some properties such as `SupportedSyntaxes` and `DefaultSyntax` that I see being used in the designer. Is there a reason I can't access these? I'm using Elsa 1.5.4.1816 from NuGet. I'm assuming they are there somewhere since they show up in the designer on the pre-made activities

**Reply by tomy2105** [reply]
Sorry, I was referring to Elsa 2.0.
Unfortunately I've only worked with Elsa 2.0 so cannot comment on Elsa 1.

**Reply by Aoshua** [reply]
It's all good. I can't wait for 2.0 to be released!

**Reply by Aoshua** [reply]
@tomy2105  Now that 2.0 is released I'm dabbling in this again and having a little trouble. I have defined some output from my first activity and want to use it in my second activity. Where do I find that output in my second activity to use it as an in-argument? Below is my code.
My workflow:
```
public class DemoWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .WriteLine("Running the demo workflow.")
            .Then<TestActivity1>(activity => activity.Set(x => x.Message, "Here is my custom message.")) // This is how you pass in a property
            .Then<TestActivity2>()
            .WriteLine("Done");
    }
}
```
Activities 1 & 2:
```
[Activity(DisplayName = "Test Activity 1", Description = "Testing...")]
public class TestActivity1 : Activity
{
    [ActivityProperty(Hint = "The message to write.")]
    public string Message
    {
        get => GetState<string>();
        set => SetState(value);
    }

    protected override IActivityExecutionResult OnExecute()
    {
        Console.WriteLine($"Executing TestActivity1: {Message}");
        var outputObject = new { Name = "Victor Hugo", Age = "83" };
        return Outcome("Done", outputObject);
    }
}
```

```
public class TestActivity2 : Activity
{
    protected override IActivityExecutionResult OnExecute()
    {
        Console.WriteLine("Executing TestActivity2");
        var inputFromPrevActivity = this.Data; // This is empty. Where else should I look?
        return Outcome("Done");
    }
}
```

**Reply by sfmskywalker** [reply]
Try this:

```csharp
public class DemoWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .WriteLine("Running the demo workflow.")
            .Then<TestActivity1>(activity => activity.Set(x => x.Message, "Here is my custom message.")) // This is how you pass in a property
            .Then<TestActivity2>(activity => activity.Set(x => x.Message, context => JObject.FromObject(context.Input))) // Using JObject because the output from activity 1 is an anonymous object.
            .WriteLine("Done");
    }
}

[Activity(DisplayName = "Test Activity 1", Description = "Testing...")]
public class TestActivity1 : Activity
{
    [ActivityProperty(Hint = "The message to write.")]
    public string Message { get; set; } // Notice that you no longer need to do `SetState` as in Elsa 1.

    protected override IActivityExecutionResult OnExecute()
    {
        Console.WriteLine($"Executing TestActivity1: {Message}");
        var outputObject = new { Name = "Victor Hugo", Age = "83" };
        return Done(outputObject); // "Done" is a convenient "alias" of "Outcome("Done", outputObject)"
    }
}

public class TestActivity2 : Activity
{
    [ActivityProperty(Hint = "The message to write.")]
    public string Message { get; set; }

    protected override IActivityExecutionResult OnExecute()
    {
        Console.WriteLine("Executing TestActivity2");
        var inputFromPrevActivity = Message; 
        return Done();
    }
}
```

As you can see, the TestActivity2 now also exposes a property that you can set from the workflow. But here's the kicker: the value that you provide from the workflow doesn't have to be "static" value that exists at the time of the workflow builder executing the workflow builder API - you get to specify lambdas that receive the current `ActivityExecutionContext` so that you can access things like the output from the previously executed activity (via the context's `Input` property), workflow variables, output from other activities, or anything really - even async calls made to other APIs if you wanted to:

` .Then<TestActivity2>(activity => activity.Set(x => x.Message, context => JObject.FromObject(context.Input)))`

**Reply by Aoshua** [reply]
@sfmskywalker, this works great, thanks!

**Reply by Aoshua** [reply]
Now I'm wondering: what would the JSON structure for a workflow that has in and out arguments look like? Can such a workflow be built with Elsa 2 Designer (still trying to get that up and running)?

**Reply by sfmskywalker** [reply]
Absolutely. The biggest difference will be that instead of writing C# lambda statements as workflow expressions, you’ll be writing them in JavaScript or Liquid, depending on the most appropriate use case. Liquid is a templating syntax convenient to use for writing things like HTTP responses and email bodies.

I’m curious to hear more about your struggles to setting up the dashboard.

I’ve written a QuickStart here: https://elsa-workflows.github.io/elsa-core/docs/next/quickstarts/quickstarts-aspnetcore-server-dashboard-and-api-endpoints

If you already went through that, then perhaps you could share with me where you got stuck / what’s unclear etc?

**Reply by Aoshua** [reply]
I haven't seen this tutorial yet, but it looks really helpful. Thanks, I'll let you know if I still have any questions afterward 😁

**Reply by Aoshua** [reply]
Yes! That was massively helpful!

**Reply by Aoshua** [reply]
> Absolutely. The biggest difference will be that instead of writing C# lambda statements as workflow expressions, you’ll be writing them in JavaScript or Liquid, depending on the most appropriate use case.

So in the designer, I have Activity1 which has some output and I have Activity2 that takes some input (with the same type as the output from Activity1). The input property on Activity2 has an `ActivityProperty` attribute setting the `DefaultSyntax = SyntaxNames.JavaScript`, but I notice that back in the designer the input is not rendered with Monaco editor. What would I type in the HTML input for this input property, if I was using JavaScript, to get the value from the previous activity? Is there a global JavaScript property that holds the value? Maybe something like `context.Input`?

**Reply by sfmskywalker** [reply]
Yeah so the DefaultSyntax is a little bit confusing,  it right now it doesn’t affect the default display mode of an activity property editor.

Just make sure that you also supply a list of SupportedSyntaxes so that the UI will give you the option of switching to the JavaScript editor, where you can simply use the `input` variable.

I’ve done a little bit of work already on support JavaScript expressions here: https://elsa-workflows.github.io/elsa-core/docs/next/expressions/expressions-javascript#input

**Reply by Aoshua** [reply]
Excellent. Thanks again for the help!

**Comment by Aoshua** [comment, marked as answer]
For now I will try to get 2.0 running in my project... But does anyone know how to achieve in and out arguments in Elsa 1.x?

**Reply by sfmskywalker** [reply]
> Output.SetVariable("someKey", myObject);
> // This at least compiles... Could this be it? (about to go test)

For Elsa 1, that's exactly what you are looking for.
Output there is a dictionary that you can access from the next activity (or any subsequent activities).

For Elsa 2, an activity can currently only return a single output (which can be any type of object as @tomy2105 mentioned), but your comment about Windows WF out-arguments reminded me of that feature, and how it would fit beautifully with Elsa as well. In fact, it would automatically provide a convenient way for the designer to infer what types of output an activity provides, which helps with things like intellisense and output value pickers.

**Reply by sfmskywalker** [reply]
I created [an issue](https://github.com/elsa-workflows/elsa-core/issues/896) for this.

**Reply by Aoshua** [reply]
Thanks! That will be a great feature to add to 2.0
